pkg/binlog/binlog.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binlog
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/pingcap/errors"
+)
+
+// Client is the client of binlog.
+type Client struct {
+	tls        *tls.Config
+	httpClient *http.Client
+	etcdClient *clientv3.Client
+
+	// if setted, will call HookAddr to change the address of pump/drainer
+	// before accessing pump/drainer.
+	HookAddr func(addr string) (changedAddr string)
+}
+
+func (c *Client) hookAddr(addr string) string {
+	if c.HookAddr != nil {
+		return c.HookAddr(addr)
+	}
+
+	return addr
+}
+
+// NewBinlogClient create a Client.
+func NewBinlogClient(pdEndpoint []string, tlsConfig *tls.Config) (*Client, error) {
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   pdEndpoint,
+		DialTimeout: time.Second * 5,
+		TLS:         tlsConfig,
+	})
+	if err != nil {
+		return nil, errors.AddStack(err)
+	}
+
+	return &Client{
+		tls: tlsConfig,
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+			},
+		},
+		etcdClient: etcdClient,
+	}, nil
+}
+
+// Close the client.
+func (c *Client) Close() error {
+	return c.etcdClient.Close()
+}
+
+func (c *Client) getURL(addr string) string {
+	scheme := "http"
+	if c.tls != nil {
+		scheme = "https"
+	}
+
+	return fmt.Sprintf("%s://%s", scheme, addr)
+}
+
+func (c *Client) getOfflineURL(addr string, nodeID string) string {
+	return fmt.Sprintf("%s/state/%s/close", c.getURL(addr), nodeID)
+}
+
+// StatusResp represents the response of status api.
+type StatusResp struct {
+	Code    int    `json:"code"`
+	Message string `json:"message"`
+}
+
+// NodeStatus represents the status saved in etcd.
+type NodeStatus struct {
```

Comment:
Is this duplicate with the one defined in `types.go`?

Reply:
the one in `types.go` us unused, i will delete it.